Make the EVP API available from within the FIPS module

mattcaswell opened the pull request:
The FIPS module will itself need to be able to use the EVP API internally. For example:

- to enable Self tests
- RAND code needs to access ciphers, HMAC etc
- DigestSign* operations will require access to digests

This makes the necessary changes to enable some limited EVP calls from within the module itself. Specifically we enable the EVP_Digest* and EVP_Cipher* functions (although currently the only algorithm available in the FIPS module is SHA256). To demonstrate this working I've included some dummy EVP calls. These are for demonstration purposes only and will eventually be removed.

This is marked as WIP because it uses #8700 as a starting point. Only the last 2 commits are relevant to this PR.

richsalz commented on crypto/context.c:
```diff
@@ -32,15 +32,18 @@ static int context_deinit(OPENSSL_CTX *ctx)
-        && context_init(&default_context)
-        && OPENSSL_atexit(do_default_context_deinit);
+    return context_init(&default_context)
+#ifndef FIPS_MODE
```
Breaking a statement in the middle is kinda yucky.  Might not be worth fixing tho.

mattcaswell replied:
This change was done in the first commit in this PR and then changed again in the second commit, so it no longer breaks in the middle.

richsalz commented on crypto/objects/search.c:
```diff
@@ -0,0 +1,65 @@
+
+#include <openssl/objects.h>
+
+const void *OBJ_bsearch_(const void *key, const void *base, int num, int size,
```
I don't recall seeing trailing underscore; isn't there another in-use convention you could use?

mattcaswell replied:
Moving this code around turned out to be unnecessary so I've removed it from the scope of this PR anyway.

richsalz commented on crypto/objects/search.c:
```diff
@@ -0,0 +1,65 @@
+#ifdef CHARSET_EBCDIC
+    /*
+     * THIS IS A KLUDGE - Because the *_obj is sorted in ASCII order, and I
+     * don't have perl (yet), we revert to a *LINEAR* search when the object
```
Remove the "I don't have perl (yet)" phrase.

mattcaswell replied:
As above (no longer in this PR)

richsalz commented on crypto/provider_core.c:
```diff
@@ -134,8 +134,7 @@ static const OPENSSL_CTX_METHOD provider_store_method = {
 {
-    return OPENSSL_init_crypto(0, NULL)
-        && (provider_store_index =
+    return (provider_store_index =
```
I think it's more clear to break this into assignment and ``` return provider_store_index != -1```

mattcaswell replied:
Done

levitte commented on providers/common/ciphers/block.c:
```diff
@@ -0,0 +1,112 @@
+    size_t i, pad = blocksize - *buflen;
+
+    for (i = *buflen; i < blocksize; i++)
+        buf[i] = pad;
```
The implicit `size_t` to `unsigned char` conversion leaves some compilers unhappy...

mattcaswell replied:
This was fixed in #8700

levitte commented on build.info:
```diff
@@ -3,7 +3,7 @@
 
 LIBS=libcrypto libssl
-INCLUDE[libcrypto]=. crypto/include include
+INCLUDE[libcrypto]=. crypto/include include providers/common/include
```
I was unsure if I should comment on this...  but I'm wondering if this shouldn't be kept with the default provider build.info.  My reasoning is that if we decide to make it possible to make the default provider a module as well some time in the future, we shouldn't need to make a default provider conditional here, or to chase around the build.info files to find this particular include setting.

mattcaswell replied:
This was actually implemented by #8700, but I've added a new commit in this PR to change it as you suggest.

levitte commented on crypto/evp/digest.c:
```diff
@@ -55,9 +55,13 @@ int EVP_MD_CTX_reset(EVP_MD_CTX *ctx)
+
+#if !defined(OPENSSL_NO_ENGINE) && !defined(FIPS_MODE)
     ENGINE_finish(ctx->engine);
 #endif
```
Quite frankly, I wonder if I don't prefer this construct

``` C
#ifndef FIPS_MODE
    /* TODO(3.0): Temporarily no support for EVP_DigestSign* in FIPS module */
    if (!EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_KEEP_PKEY_CTX))
        EVP_PKEY_CTX_free(ctx->pctx);

# ifndef OPENSSL_NO_ENGINE
    ENGINE_finish(ctx->engine);
# endif
#endif
```

That will makes things easier to parse for the upcoming source integrity checker.
```

mattcaswell replied:
Done

slontis commented on crypto/evp/evp_lib.c:
```diff
@@ -296,6 +360,31 @@ int EVP_CIPHER_CTX_nid(const EVP_CIPHER_CTX *ctx)
+
+        /* Cipher comes from a provider - so ask the provider for the mode */
+        const OSSL_PARAM params[] = {
+            OSSL_PARAM_int(OSSL_CIPHER_PARAM_MODE, &mode),
```
I think you need to use this instead.
OSSL_PARAM params[2];
size_t n = 0;

params[n++]  = OSSL_PARAM_construct_int(..);
params[n++] = OSSL_PARAM_construct_end();

mattcaswell replied:
This was fixed in #8700

slontis commented on crypto/provider_core.c:
```diff
@@ -560,13 +575,21 @@ const DSO *ossl_provider_dso(OSSL_PROVIDER *prov)
 const char *ossl_provider_module_path(OSSL_PROVIDER *prov)
 {
+#ifdef FIPS_MODE
+    return NULL;
```
Why do you do this?

mattcaswell replied:
To avoid bringing the DSO stuff inside the module boundary. It's not needed.

slontis commented on include/openssl/core_numbers.h:
```diff
@@ -78,14 +85,15 @@ OSSL_CORE_MAKE_FUNC(const OSSL_ALGORITHM *,provider_query_operation,
 # define OSSL_FUNC_DIGEST_NEWCTX            1
 # define OSSL_FUNC_DIGEST_INIT              2
-# define OSSL_FUNC_DIGEST_UPDDATE           3
+# define OSSL_FUNC_DIGEST_UPDATE            3
```
This is also modified in the digest PR.

slontis commented on providers/common/ciphers/aes.c:
```diff
@@ -0,0 +1,402 @@
+    const OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_PADDING);
+    if (p != NULL && !OSSL_PARAM_set_uint(p, ctx->pad))
```
why do you use a uint here and a int further down in the set_params?

mattcaswell replied:
This was fixed in #8700

slontis commented:
I assume at some point we may want to select properties such as "asm=no" or "asm=yes" Does this fit is well with the capability tests??... or should this be handled differently. i.e 2 dispatch entries (one for the generic version + others for asm version(s))

mattcaswell commented:
> I assume at some point we may want to select properties such as "asm=no" or "asm=yes" Does this fit is well with the capability tests??... or should this be handled differently. i.e 2 dispatch entries (one for the generic version + others for asm version(s))

I'm not sure that is actually desirable. But if we decide to do it then we can split the dispatch table entries at that point.

mattcaswell commented:
Now that #8700 has gone in I have rebased this and addressed all comments received so far.

mattcaswell commented:
Not sure why AppVeyor is claiming this is non-mergeable with the master branch because this is a fresh rebase, Travis has no such problems, and github reports no merge conflicts.

levitte commented on crypto/asn1/a_object.c:
```diff
@@ -18,6 +18,8 @@
 #include "asn1_locl.h"
 
+#if !defined(FIPS_MODE)
+
```
nit: empty line unnecessary

mattcaswell replied:
Fixed.

levitte commented:
I look at this, and gotta say that I'm very very doubtful about bringing the library context stuff into the FIPS provider (or really, any provider).

levitte commented:
> I'm very very doubtful about bringing the library context stuff into the FIPS provider

Note that it's not like I understand why it gets dragged in...  I think I need a little bit of time to digest this.

levitte commented on providers/fips/fipsprov.c:
```diff
@@ -88,12 +146,84 @@ int OSSL_provider_init(const OSSL_PROVIDER *provider,
+    if (!default_method_store_init()) {
+        fips_teardown();
+        return 0;
+    }
```
So....  instead of all of that, could we simply have a provider module global library context, i.e. use OPENSSL_CTX_new() to create it and OPENSSL_CTX_free() to free it, and use it everywhere we can?  It seems like a saner approach than doing all kinds of internal calls, with all the entanglement that could result in in the future.

mattcaswell replied:
Yes, I think that would be nice. We should use OPENSSL_CTX wherever we have global state we need to initialise. But I see that as outside of the scope of this PR.

levitte replied:
I don't.  Basically, you've made a load of static initialization functions more widely available, which has a risk of entanglement I already see with current init functionality.  You could achieve the exact same thing by simply creating a library context within the FIPS provider, and by freeing it on teardown.  No need to go call all those internal initializers from outside (or should not be any need at the very least).

Since you do already include context.c in the FIPS module, you already have all you need to do this, so leaving it for another PR will basically mean undoing the bunch of init stuff you've entangled yourself with here.  Therefore, I'd say this is very much *in* scope.

mattcaswell commented:
> I look at this, and gotta say that I'm very very doubtful about bringing the library context stuff into the FIPS provider (or really, any provider).

This PR brings EVP inside the FIPS boundary. EVP uses core in order to fetch algorithms and all of that code depends on OPENSSL_CTX. So if you want EVP inside the boundary, you must have OPENSSL_CTX.

mattcaswell commented:
I pushed a fixup commit addressing the comment above. I also took this out of WIP.

richsalz commented:
> So if you want EVP inside the boundary, you must have OPENSSL_CTX.

Unless you only pass NULL, right?

I'd go further and remove the library context stuff for this release (I forget which PR I made that comment) and only allow NULL and the global context. As I said "over there," I think the context stuff isn't fully thought out yet.

mattcaswell commented:
>> So if you want EVP inside the boundary, you must have OPENSSL_CTX.
>
> Unless you only pass NULL, right?

You misunderstand my statement. I am saying that in order for the EVP code to be inside the FIPS boundary, there is a dependency on the OPENSSL_CTX code. This is nothing to do with passing NULL for the default context or not.

richsalz commented:
Yes, I misunderstood, thanks.  I wonder if we should ask the lab for a viewpoint on this?

levitte commented:
Hmmm...  assuming that there is just the default library context and none else means that we can ignore issues with multiple initialization (i.e. the RUN_ONCE problem that we're seeing here).  That would be @richsalz' point, yeah?

I'm not too keen on ignoring the library context thing, but I have to admit it's a line of thinking worth spending more than just a fleeting thought on.

mattcaswell commented:
I've updated this to make the changes suggested above and I've pulled in OPENSSL_atexit() and OPENSSL_cleanup() into the FIPS module. The ex_data stuff is really properly handled yet. That will take #8828 to go in first.

richsalz commented:
I don't understand why atext and cleanup are pulled in.  If there is module-specific "cleanup on unload" code, then the core should query for a function pointer and call into the module once it is unloaded. No shared-library module is loaded before the OpenSSL core is initialized. If an application calls init without a flag that the module needs, then loading the module can signal an error; the app is broken for the module it is trying to use.

mattcaswell commented:
> If there is module-specific "cleanup on unload" code, then the core should query for a function pointer and call into the module once it is unloaded.

This is exactly what does happen. And the module specific function that is called by the core to cleanup on unload is called....OPENSSL_cleanup(). The objective is to share as much code as possible between libcrypto and the FIPS module. We do not want to write two versions of code for everything: once for libcrypto and once for the FIPS module. That has been our approach all the way through the design process. This approach applies to initialisation and cleanup too.

richsalz commented:
The module-specific cleanup code shouldn't need anything that overlaps with the core. Everything a module needs to do that's similar should be implemented via callbacks into the core. As I wrote elsewhere, the exdata seems like the first, major, example of this.

mattcaswell commented:
> The module-specific cleanup code shouldn't need anything that overlaps with the core. Everything a module needs to do that's similar should be implemented via callbacks into the core

That is generally true for providers. It is not true for the FIPS module. Apart from some minor exceptions where the lab said it was ok (such as for error handling) the whole design for the FIPS module is that it is self contained. If we didn't need to do that then we would just keep the whole of the EVP code in core and not have to worry about bringing it into the FIPS module at all. As soon as you bring the EVP code in (which has been in the design since the very beginning) then you also have to bring in all its dependencies. This includes ex_data. If we don't do this then the FIPS module will not be able to (for example) do a DigestSign operation without the help of code outside the boundary, or (as another example) do its own self tests without the help of external code.

richsalz commented:
I understand bringing in EVP. Saying that means you have therefore bring in all dependencies isn't necessarily true -- as I recall from the last F2F, the question of memory allocation and secure heap were still open. I also don't recall a decision to make the FIPS module "self contained," and if I'm wrong, I am now saying we should reconsider that because of the complexity that we are starting to see with this and the #8828 and the problems I am convinced it will cause, we should confirm if the simpler approach is acceptable.

levitte commented on providers/fips/fipsprov.c:
```diff
@@ -88,12 +139,71 @@ int OSSL_provider_init(const OSSL_PROVIDER *provider,
+        OPENSSL_cleanup();
+        return 0;
+    }
+
```
This section is still problematic.  There should be no reason to do all the manual initialization here, since the affected stores (should) self-initialize on demand.

Also, I'm surprised to not see a teardown function calling `OPENSSL_cleanup`, I though that would be a fairly obvious thing to do.

Finally, I understand that you worry about multiple init of this module, due to the possibility of multiple library contexts.  I'm not sure how this helps, though (with the lack of a teardown that cleans up), but also, the experience with #8828 could suggest that an init counter would be useful, but here in `OSSL_provider_init` and associated teardown function rather than in the ex_data code.

mattcaswell replied:
> This section is still problematic. There should be no reason to do all the manual initialization here, since the affected stores (should) self-initialize on demand.

We have no mechanism to enable that at the moment since RUN_ONCE just doesn't work in the general case. It could be useful where no actual resources are allocated that need to be later freed (such as where we just assign an ex_data index), but if we allocate resources we're a bit stuck. Even the ref counter doesn't work in that case since we cannot guarantee that if the ref counter goes to zero that the .bss and .data will be reinitialised before the provider is loaded again.

> Also, I'm surprised to not see a teardown function calling OPENSSL_cleanup, I though that would be a fairly obvious thing to do.

Since the only thing the teardown would need to do is call OPENSSL_cleanup, and the teardown function and OPENSSL_cleanup have the same signature, I just used OPENSSL_cleanup directly as the teardown function:

```
static const OSSL_DISPATCH fips_dispatch_table[] = {
    { OSSL_FUNC_PROVIDER_TEARDOWN, (void (*)(void))OPENSSL_cleanup },
```
I could insert an intermediate teardown, but it seems to add little value.

> Finally, I understand that you worry about multiple init of this module, due to the possibility of multiple library contexts. I'm not sure how this helps, though (with the lack of a teardown that cleans up), but also, the experience with #8828 could suggest that an init counter would be useful, but here in OSSL_provider_init and associated teardown function rather than in the ex_data code.

Yes there is still an issue here. As we discussed offline last week I think we need to have OSSL_provider_init return a provider side handle that corresponds to the libcrypto side OPENSSL_CTX object.

levitte replied:
[sigh]

The elepahnt in the room is that we're doing everything we can to avoid touching shared object init/fini type of constructors/destructors.  The more I see us doing all kinds of not very elegant stuff to avoid that, the more I'm wondering if we're not making a mistake.  This discussion is pretty indicative.

mattcaswell replied:
From the "Self Testing" section of the design document:

"The module's Default Entry Point (DEP), the ".init" function in Linux libraries, will set a module variable (probably the state variable). This variable will be checked in the OSSL_provider_init() and, if set (which it always will be), will verify the values in the file. This two-step process meets the FIPS requirement of having the DEP ensure the tests are run, but allows us to implement the tests when the rest of the module is being initialized during normal operation."

So we have already recognised the need for a ".init" in the FIPS module.

paulidale replied:
FIPS requires the .init constructor or equivalent (if available).  The note from the standard (whose name escapes me) even includes sample code for Linux/Unix and Windows.  A FIPS provider wouldn't be limited to not doing more in its constructor, so long as it doesn't offer any services before finishing its self tests.

Is it too much of a stretch going from _providers are shared libraries_ to _providers are shared libraries with initialisers_?

The alternative seems to be providers can be loaded but never unloaded and there might be restrictions on when they can be loaded.

mattcaswell commented:
See #8857 for a solution to the self-initialisation problem.

mattcaswell commented:
I've rebased this now that #8857 has gone in. The result of that PR is that none of the controversial manual initialisation and cleanup code is required anymore in this PR, so it has all gone. @levitte - please take a look.

paulidale commented on crypto/asn1/a_object.c:
```diff
@@ -18,6 +18,7 @@
 #include "internal/asn1_int.h"
 #include "asn1_locl.h"
 
+#if !defined(FIPS_MODE)
```
We _really_ don't want _any_ of the ASN1 code inside the FIPS boundary.  Not even including the internal headers.  Any necessary ASN1 work needs to be done via callbacks or hardwired.  The latter is the usual way.

levitte replied:
Except that `ASN1_OBJECT` is the fundamental type to store the data associated with the NIDs created with the routines in `crypto/objects/`...  Of course, we could use an entirely different NID ↔ name database within the FIPS module, but that's a bigger effort and double maintenance of similar features.

paulidale replied:
Is a NID↔name database even needed inside the FIPS boundary?  Internally, it will use NIDs won't it?  The name resolution should happen outside of the boundary, it is not a cryptographic operation.

If the database is required, would it be possible to treat `ASN1_OBJECT` as opaque pointer and use callbacks?

I agree that a parallel database isn't the way to go.

mattcaswell replied:
> The name resolution should happen outside of the boundary, it is not a cryptographic operation.

There will be EVP_XX_fetch() calls from within the boundary. For example in the self tests. This requires name to nid resolution.

> We really don't want any of the ASN1 code inside the FIPS boundary.

For all practical purposes there isn't any. We're not bringing in any code that generates or parses ASN.1. The *only* code we are bringing in here is the ASN1_OBJECT_new() function and the ASN1_OBJECT_free() function. Nothing else. I don't really see why that should be considered a problem.

paulidale replied:
Good point about the fetches.

The new and free aren't the main concern.  The internal ASN1_OBJECT structure is #included which means that entire header is inside the boundary. Splitting the header might help as would creating accessor functions.

FIPS only has a couple of dozen algorithms in total and only a subset of these will be referenced by the self tests, perhaps a parallel name↔NID function is workable (thinking a string of `if strcmp else if strcmp` rather than anything fancy).

paulidale replied:
@slontis do you have the list of power up self tests that are necessary?

slontis replied:
design doc is the closest we have to a list..
Its not just the self tests though right? its anything that contains multiple layers of algs (like HMAC)

paulidale replied:
Agreed.  I figured the self test list would be a beginning towards what names would need to be mapped.

HMAC doesn't necessarily have to resolve the digest name -- it could take a NID directly.  This will depend on the OSSL_PARAM definition that is finally used -- either NID or name could be passed through.  I prefer not exposing NIDs to users but providers will need to know about them.

t-j-h replied:
I suggest that we split the NID handling ASN1 APIs into a new OSSL_NID api set and leave behind wrapper functions.
Disconnect the mapping logic from the ASN1 logic itself - they are independent - so we should make the APIs show it. We should not have any ASN1 headers used in the providers - its too easy to make mistakes like that.

And this does demonstrate that we should split out the NID handling into a separate namespace as such.

It will look a whole lot cleaner if that is done IMHO.

t-j-h replied:
And users shouldn't see NIDs in the API at all ... so where do we actually need that logic?
i.e. if we need it then split it out, but why do we need it at all in a provider?

levitte replied:
It should certainly be possible to sseparate the NID ↔ name map from the NID ↔ OID map.  The ASN1_OBJECT lumps them all together, but that's not really necessary.

levitte replied:
As a matter of fact, all we need is a number ↔ name map in `crypto/evp/evp_fetch.c`, which doesn't anything fancier than a LHASH indexed by name and with a \< number, name \> tuple as value.  The numbers we use in `evp_fetch.c` can all be created on the fly, it's not like the numbers themselves carry any more meaning than being a current index into a dynamic database.

levitte replied:
#8878 is my attempt to get away from using the ASN1_OBJECT store.  Note that it's currently a draft, I have seen it build, but have made no test or documentation yet...  To be finished next week if that PR is acceptable.

paulidale replied:
Is it prudent to introduce a _new_ set of integers to map algorithm names too?  That's what NIDs are doing isn't it?

The only advantage, I can see, of having another integer is that they then can't be used as NIDs.  I'm fine with & would encourage this but I don't think it makes sense to complicate the code to get it.

If we insist on on using different numbers, they can be converted algorithmically (e.g. the NID * day-of-month the library was built) rather than having a map.

levitte replied:
This was the intent from the very start when I made the constructors.  The numeric identities are to be seen as session data rather than constants.  The *only* reason I dealt with numeric identities at all was that the method store uses numeric identities as input rather than names.

kaduk replied:
> I prefer not exposing NIDs to users but providers will need to know about them.

I'm a bit confused about why even providers need to know about NIDs.

> The numeric identities are to be seen as session data rather than constants

E.g., in this mindset we only have ephemeral name/number mappings (which could potentially even be unique per provider, if we want)

paulidale replied:
@kaduk you are right, providers shouldn't need to know about NIDs.

levitte commented on include/openssl/core_numbers.h:
```diff
@@ -57,6 +58,12 @@ OSSL_CORE_MAKE_FUNC(const OSSL_ITEM *,
+                                         const char *file, int line))
+# define OSSL_FUNC_CORE_ADD_ERROR_VDATA        4
+OSSL_CORE_MAKE_FUNC(void,core_add_error_vdata,(int num, va_list args))
+
```
I've been staring at this many times, thinking something is missing...  and yeah, you want upcalls for `ERR_func_error_string` and `ERR_load_strings_const` as well, for the sake of `ERR_load_PROV_strings`

richsalz replied:
I think that can go in a separate PR if needed.  I don't think the FIPS provider will add error codes that aren't in the core, and having providers add their own error codes is a bigger problem that would "distract" from the main point of this.

mattcaswell replied:
> I've been staring at this many times, thinking something is missing... and yeah, you want upcalls for ERR_func_error_string and ERR_load_strings_const as well, for the sake of ERR_load_PROV_strings

Why do we need that? Error string are all libcrypto side - why do we need to expose them to the provider?

levitte replied:
Uhmmm, it's the other way around.  Provider side error messages should be in the provider, and the means that the provider will need to have them loaded so libcrypto can display them right.

richsalz replied:
Providers cannot load their own error messages until OpenSSL defines a way for an arbitrary third-party to get a library/facility that doesn't conflict with any other. As the number of bits available is very very small, solving this will be a hard problem.

levitte replied:
The number of bits is 8, allowing up to 256 different error "modules". There is actually a function to obtain the next available number: `ERR_get_next_error_library`. That one would need an upcall too. 

Note that this is with our current error model. That one could be remade to become much simpler to handle.

mattcaswell replied:
> Uhmmm, it's the other way around. Provider side error messages should be in the provider, and the means that the provider will need to have them loaded so libcrypto can display them right.

We need a new error system for the providers. We need to give some thought into how that will work. This PR isn't the right location to have that discussion. All I'm doing here is the minimum necessary to get this to work for now. It is not a long term solution.

I fully expect the long term solution to have errors on the provider side. I also expect that these errors will work differently to the way the current error system works. But we are not there yet. Right now we are using the existing error system. Right now error strings are loaded on the libcrypto side. Exposing upcalls for `ERR_func_error_string`, `ERR_load_strings_const`and `ERR_load_PROV_strings` is not needed for this PR. Because I expect the new error system to work differently to the current system I do not believe we should be anticipating what upcalls we are going to need for that new error system in this PR. 

I'm only exposing this particular upcall because it is necessary to get things to work right now. I do expect this to change.

levitte replied:
I can accept that.

richsalz replied:
Another option, *for this release,* is to not solve this problem, and instead do what I provide in #8887

levitte replied:
#8887, in *addition* to what is done in this PR, makes `ERR_add_error_vdata` append strings instead of replacing (which I think is a good thing, don't get me wrong). That could be an independent PR.

richsalz replied:
Yes, that is what I meant by "how I addressed the issue." :)  If there are concerns about that change, I will make a new API to append a text string. But that should be discussed over in #8887.

This PR needs the set-error API's, and @mattcaswell didn't want to delay and rebase again for 8887 to go in first.

paulidale commented:
I don't think the FIPS provider needs to specify properties on its internal queries.  Leaving out the property code makes sense.

I'm less confident about `crypto/provider_core.c`, to allow both forwards and backward compatibility for FIPS provider modules we can't freeze anything which could be a risk with shims.

levitte commented:
> I'm less confident about `crypto/provider_core.c`, to allow both forwards and backward compatibility for FIPS provider modules we can't freeze anything which could be a risk with shims.

Not sure where the compatibility concern would be, since the internal provider will always be itself!

Quick analysis...  we need:
- `ossl_provider_query_operation`.  That could easily be shimmed to simply execute `fips_query`.
- `ossl_provider_ctx`.  A simple module local `OSSL_PROVIDER` structure that only holds a `void *provctx` would make a straight copy of that function from `provider_core.c` both possible and sane.

Finally, this simple (untested) patch should show where I see this going:

``` diff
diff --git a/crypto/core_fetch.c b/crypto/core_fetch.c
index 6c4ed6a30a..c38c3f804f 100644
--- a/crypto/core_fetch.c
+++ b/crypto/core_fetch.c
@@ -94,8 +94,13 @@ void *ossl_method_construct(OPENSSL_CTX *libctx, int operation_id,
         cbdata.force_store = force_store;
         cbdata.mcm = mcm;
         cbdata.mcm_data = mcm_data;
+#ifdef FIPS_MODE
+        /* The FIPS module only has one provider: itself */
+        ossl_method_construct_this(fips_self(libctx), &cbdata);
+#else
         ossl_provider_forall_loaded(libctx, ossl_method_construct_this,
                                     &cbdata);
+#endif
 
         method = mcm->get(libctx, cbdata.store, propquery, mcm_data);
         mcm->dealloc_tmp_store(cbdata.store);
```

A line of thought behind this is that I do not think there is a need to re-run initialization for recursive uses of the FIPS module.

paulidale commented:
The concern is that we want to be able to run the provider with future versions of OpenSSL (e.g. FIPS 3.0 with OpenSSL 3.1 definitely & 4.0 maybe) and vice versa (e.g. OpenSSL 3.0 using the FIPS 3.1). Although the latter case would be less common and potentially could be ruled out.  From the explanation given, it seems like this shouldn't be a problem but by special casing things, there is a risk that it becomes so in the future.

levitte commented:
A question that came up while investigating the ramifications of #8878...  will implicit fetches ever happen in the FIPS module, or can we somehow avoid that?

levitte commented:
> The concern is that we want to be able to run the provider with future versions of OpenSSL (e.g. FIPS 3.0 with OpenSSL 3.1 definitely & 4.0 maybe) and vice versa (e.g. OpenSSL 3.0 using the FIPS 3.1). Although the latter case would be less common and potentially could be ruled out. From the explanation given, it seems like this shouldn't be a problem but by special casing things, there is a risk that it becomes so in the future.

I'm not sure I follow...  The stuff I'm discussing is *internal to the FIPS module*!  I don't see that slipping out, or if it does, we're doing something fundamentally wrong.

levitte commented:
Regarding the dependency on OBJ and thereby potentially ASN1, see #8899

mattcaswell commented:
> A question that came up while investigating the ramifications of #8878... will implicit fetches ever happen in the FIPS module, or can we somehow avoid that?

There will never be implicit fetches in the FIPS module.

levitte commented:
#8878 has been merged, this PR will need to be rebase and possibly adapted.

mattcaswell commented:
I rebased this now that #8878 has gone in. There are no longer any dependencies on ASN.1 code.

slontis commented:
NOTE: the new namemap stuff introduces an issue with aliases. The old maps had both the short name and long names registered with the NID. We would now need a duplicated name for every value in the provider tables.

levitte commented:
The new namemap doesn't introduce said issue with aliases, the issue was always there!  The only reason we didn't notice was that we piggy-backed on the object database, but as soon as a provider offered algorithm names we don't know beforehand, there are no aliases in sight.

I remember that we talked about aliases at some point, but we never concluded anything that I can recall.

paulidale commented:
Would an alias API make sense here?  The numeric value could map to the first string registered and all string aliases could map to the same numeric value.

The obvious alternatives would be to:
- use NIDs directly (which is being avoided here) or
- let the provider deal with it (which isn't a good API)

levitte commented:
On the recursive provider activation:

> > Finally, this simple (untested) patch should show where I see this going:
> 
> I think this will fail to activate the recursive FIPS provider on first use. The recursive FIPS provider is flagged as a fallback, and since we don't explicitly load any other providers the fallback code will kick in.

Well yeah, it simply uses itself (`fips_self(libctx)`, which is currently unimplemented).

> I am sure there are some things we could do in the core code to "short circuit" everything given that we know that there will only ever be one provider. But in doing so we add additional "ifdef" complexity to the code. It seems to me that where possible we should be avoiding "special casing" the FIPS module code. Where possible we just do it the same way in FIPS that we do it elsewhere.

Come again? I want to have *one* `#ifdef FIPS_MODE` in `core_fetch.c` rather than a number of them in `provider_core.c` (I don't want `provider_core.c` to be involved at all).

> > A line of thought behind this is that I do not think there is a need to re-run initialization for recursive uses of the FIPS module.
> 
> Initialisation is virtually a no-op in the recursive case, so I don't see this as being a big deal? All it does is pass back its dispatch table.
> 
> ````
> OSSL_provider_init_fn fips_intern_provider_init;
> int fips_intern_provider_init(const OSSL_PROVIDER *provider,
>                               const OSSL_DISPATCH *in,
>                               const OSSL_DISPATCH **out,
>                               void **provctx)
> {
>     *out = intern_dispatch_table;
>     return 1;
> }
> ````

The big deal for me is 1) the recursive activation *is* added complexity, especially if you want to follow what's going on in something like a debugger, and 2) that's quite a few cycles spent on a no-op.

levitte commented:
On aliases:

The provider will have to be involved *somehow*...  who else should decide what's an alias of what?

mattcaswell commented:
> Come again? I want to have one #ifdef FIPS_MODE in core_fetch.c rather than a number of them in provider_core.c (I don't want provider_core.c to be involved at all).

But surely if we don't go into provider_core.c at all we never end up calling provider_activate(), which means we never setup the provider callback functions...most importantly "query_operation"?

paulidale commented:
I doubt that most providers will want to know about aliases for existing algorithms -- these should be setup by `libcrypto`, the default provider or the legacy one.  Other providers shouldn't need to care about aliases.

For new algorithms, a provider needs to be able to register such and add aliases but this will be the exception.

levitte commented:
> I doubt that most providers will want to know about aliases for existing algorithms -- these should be setup by `libcrypto`, the default provider or the legacy one.  Other providers shouldn't need to care about aliases.

The *providers* already have the knowledge of algorithm *names*, why should they not have the knowledge of appropriate *aliases*?  Code wise, it's not very hard to implement:

``` diff
diff --git a/include/openssl/core.h b/include/openssl/core.h
index cf4d3f41de..ea066bd4a4 100644
--- a/include/openssl/core.h
+++ b/include/openssl/core.h
@@ -61,7 +61,7 @@ struct ossl_item_st {
  * An array of these is always terminated by algorithm_name == NULL
  */
 struct ossl_algorithm_st {
-    const char *algorithm_name;      /* key */
+    const char **algorithm_names;    /* key */
     const char *property_definition; /* key */
     const OSSL_DISPATCH *implementation;
 };
```

Granted, that means that the provider will have to have a separate array of names, but that's not very hard either.  I don't believe it's very hard to add alias handling to `core_fetch.c` and `core_namemap.c`, which would finish it all off...

paulidale commented:
It also means that every provider needs to know all of the names for each algorithm.  I'd prefer the core to add the aliases it knows about and for providers to not need to know about them (although they could if they wanted to).

levitte commented:
@paulidale, I completely disagree with you on this...  but also, that's not necessarily something that needs to be resolved in *this* PR, so we should take that discussion elsewhere.

levitte commented:
On the recursive provider activation:

> But surely if we don't go into provider_core.c at all we never end up calling provider_activate(), which means we never setup the provider callback functions...most importantly "query_operation"?

Why do we need to call `provider_activate()` recursively?  Surely, the `OSSL_PROVIDER*` that is passed to `OSSL_provider_init` already has all the information we need?  All that's needed is to save that pointer, and since we already create a library context, that's easily solved...

mattcaswell commented:
> Surely, the OSSL_PROVIDER* that is passed to OSSL_provider_init already has all the information we need? All that's needed is to save that pointer, and since we already create a library context, that's easily solved...

OSSL_PROVIDER is an opaque type. The OSSL_PROVIDER object that is passed to `OSSL_provider_init` was created libcrypto side. When we make EVP calls recursively from inside the FIPS module we cannot assume that the definition of the OSSL_PROVIDER structure that the FIPS module knows about is the same definition as the one used in libcrypto. The FIPS module maybe being run with a libcrypto from a different OpenSSL version and therefore the OSSL_PROVIDER structure definitions may be different.

levitte commented:
> When we make EVP calls recursively from inside the FIPS module we cannot assume that the definition of the OSSL_PROVIDER structure that the FIPS module knows about is the same definition as the one used in libcrypto. 

Yup, I just realised that as well...  so OK, I'm backing off on this one, but would like some prominent commentary on this, 'cause believe you me, the question I raised will surely be asked again unless we provide justification.

mattcaswell commented:
> would like some prominent commentary on this

I added a new commit with some additional commentary in provider_activate to explain the recursive nature of this. I also fixed an issue where we need to set the provctx value for the recursive initialisation call to be the current library context. This will ensure that EVP calls made from within the FIPS module are always associated with the current library context.

levitte commented on crypto/provider_core.c:
```diff
@@ -342,7 +342,7 @@ static const OSSL_DISPATCH *core_dispatch; /* Define further down */
  * appropriate
  */
-static int provider_activate(OSSL_PROVIDER *prov)
+static int provider_activate(OSSL_PROVIDER *prov, OPENSSL_CTX *libctx)
```
I see this as a temporary hack, 'cause this *assumes* the provider uses a library context as a provider context...  which happens to be true for the *current* FIPS module implementation...  so we need to make sure there are notes, both here and in `fipsprov.c` that this parameter must be what the module uses, or else...

(Here be dragons)

levitte replied:
Extra commentary that goes into this noted, and appreciated

levitte commented on crypto/provider_core.c:
```diff
@@ -397,6 +397,22 @@ static int provider_activate(OSSL_PROVIDER *prov)
+     * from the FIPS module itself are available via the FIPS module EVP
+     * interface, i.e. we only ever have one provider available inside the FIPS
+     * module - the FIPS provider itself. We default the provctx value to our
+     * current library context so that EVP calls from within the module are
```
I find it a bit unclear exactly what library context is the "current" one in this context...

levitte commented on crypto/provider_core.c:
```diff
@@ -436,9 +452,9 @@ static int provider_activate(OSSL_PROVIDER *prov)
 }
 
-int ossl_provider_activate(OSSL_PROVIDER *prov)
+int ossl_provider_activate(OSSL_PROVIDER *prov, OPENSSL_CTX *libctx)
```
Do you actually use this anywhere?  I can't see it...  I would suggest that *this* function should simply pass NULL as second argument to `provider_activate`

levitte replied:
Ok, having looked around (especially at what `fipsprov.c` does, I don't see the necessity to change `ossl_provider_activate`.  I understand the `provider_activate(prov, ctx)` call in `ossl_provider_forall_loaded`, and unless I'm missing something, that's the only place where this is needed to cover for the FIPS module's need.

levitte replied:
Ok, having looked around (especially at what `fipsprov.c` does, I don't see the necessity to change `ossl_provider_activate`.  I understand the `provider_activate(prov, ctx)` call in `ossl_provider_forall_loaded`, and unless I'm missing something, that's the only place where this is needed to cover for the FIPS module's need.

levitte commented on crypto/provider.c:
```diff
@@ -21,7 +21,7 @@ OSSL_PROVIDER *OSSL_PROVIDER_load(OPENSSL_CTX *libctx, const char *name)
         return NULL;
 
-    if (!ossl_provider_activate(prov)) {
+    if (!ossl_provider_activate(prov, libctx)) {
```
Oh!  This didn't show when I looked at the commit that implements this change...

levitte replied:
... but hold on, I want to look a bit further...  I'm still not convinced this is needed specifically for the FIPS module.

mattcaswell commented:
Fixup commit pushed to addressed your latest comments. I reversed the changes to `ossl_provider_activate` and restricted them to just `provider_activate`. I also adjusted and added more commentary.

@levitte - please take another look.

mattcaswell commented:
Woop!!!! Pushed.....finally!!!

Thanks to everyone involved. We got there in the end. :-)
